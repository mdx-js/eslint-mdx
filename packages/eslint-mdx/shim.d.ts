declare module 'espree/lib/token-translator' {
  export default TokenTranslator

  // eslint-disable-next-line @typescript-eslint/consistent-type-imports
  export type acorn = typeof import('acorn')

  export interface Location {
    /**
     * The start position.
     */
    start: acorn.Position
    /**
     * The end position.
     */
    end: acorn.Position
  }

  export type Range = [number, number]

  export interface EsprimaToken {
    /**
     * The type of this token.
     */
    type: string
    /**
     * The string content of the token.
     */
    value: string
    /**
     * Location in source text.
     */
    loc: Location | undefined
    /**
     * start column.
     */
    start: number | undefined
    /**
     * end column.
     */
    end: number | undefined
    /**
     * [start, end] range
     */
    range: Range | undefined
  }

  declare function TokenTranslator(
    acornTokTypes: Record<string, acorn.TokenType>,
    code: string,
  ): void

  declare class TokenTranslator {
    _acornTokTypes: Record<string, acorn.TokenType>

    constructor(acornTokTypes: Record<string, acorn.TokenType>, code: string)

    onToken(
      token: acorn.Token,
      extra: {
        // eslint-disable-next-line no-magic-numbers, sonar/max-union-size
        ecmaVersion: 'latest' | 3 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13
        tokens: EsprimaToken[]
      },
    ): void
  }
}
