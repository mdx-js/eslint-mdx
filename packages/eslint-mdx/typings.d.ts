declare interface SyntaxError {
  column?: number
  index?: number
  lineNumber?: number
}

declare module 'espree' {
  import * as estree from 'estree'

  // The version of espree that's loaded, like "v3.4.0".
  export const version: string

  // Parse the given text as javascript into an abstract syntax tree.
  export function parse(text: string, opts?: Options): estree.Program

  // Walk the given text and produce an array of tokens.
  export function tokenize(text: string, opts?: Options): Token[]

  interface Token {
    type: string
    value: string
    start: number
    end: number
  }

  // A map from node type to the properties on that type that contain more nodes.
  export const VisitorKeys: { [kind: string]: string[] }

  const node: estree.Node
  // A map of node types to themselves.
  export const Syntax: { [type: string]: typeof node.type }

  // Options for parsing.
  interface Options {
    // attach range information to each node
    range?: boolean

    // attach line/column location information to each node
    loc?: boolean

    // create a top-level comments array containing all comments
    comment?: boolean

    // attach comments to the closest relevant node as leadingComments and
    // trailingComments
    attachComment?: boolean

    // create a top-level tokens array containing all tokens
    tokens?: true

    // set to 3, 5 (default), 6, 7, or 8 to specify the version of ECMAScript
    // syntax you want to use.
    // You can also set to 2015 (same as 6), 2016 (same as 7), or 2017 (same as 8)
    // to use the year-based naming.
    ecmaVersion?: number

    // specify which type of script you're parsing (script or module, default is script)
    sourceType?: 'script' | 'module'

    // specify additional language features
    ecmaFeatures?: {
      // enable JSX parsing
      jsx?: boolean

      // enable return in global scope
      globalReturn?: boolean

      // enable implied strict mode (if ecmaVersion >= 5)
      impliedStrict?: boolean

      // allow experimental object rest/spread
      experimentalObjectRestSpread?: boolean
    }
  }
}

declare module 'remark-mdx' {
  import * as unified from 'unified'
  const mdx: unified.Attacher
  export = mdx
}
