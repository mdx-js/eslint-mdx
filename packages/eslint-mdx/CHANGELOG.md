# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.9.6](https://github.com/rx-ts/eslint-mdx/compare/v0.9.5...v0.9.6) (2019-08-07)


### Bug Fixes

* custom parser resolved undefined error ([1bbb1ff](https://github.com/rx-ts/eslint-mdx/commit/1bbb1ff))

## [0.9.5](https://github.com/rx-ts/eslint-mdx/compare/v0.9.3...v0.9.5) (2019-08-05)


### Bug Fixes

* options unchangeable issue for default parser ([e76c483](https://github.com/rx-ts/eslint-mdx/commit/e76c483))

## [0.9.4](https://github.com/rx-ts/eslint-mdx/compare/v0.9.3...v0.9.4) (2019-08-05)


### Bug Fixes

* adjacent jsx nodes should be allowed in mdx ([8456381](https://github.com/rx-ts/eslint-mdx/commit/8456381))
* skip combine jsx nodes for root/jsx node ([90583f7](https://github.com/rx-ts/eslint-mdx/commit/90583f7))

## [0.9.3](https://github.com/rx-ts/eslint-mdx/compare/v0.9.1...v0.9.3) (2019-08-04)


### Bug Fixes

* error from typescript parser in plain object ([f634e40](https://github.com/rx-ts/eslint-mdx/commit/f634e40))

## [0.9.1](https://github.com/rx-ts/eslint-mdx/compare/v0.9.0...v0.9.1) (2019-08-03)


### Bug Fixes

* always extends base config ([530160a](https://github.com/rx-ts/eslint-mdx/commit/530160a))

## [0.9.0](https://github.com/rx-ts/eslint-mdx/compare/v0.8.1...v0.9.0) (2019-08-03)


### Features

* support other extensions, detect parsers automatically ([5386098](https://github.com/rx-ts/eslint-mdx/commit/5386098))

## [0.8.1](https://github.com/rx-ts/eslint-mdx/compare/v0.7.1...v0.8.1) (2019-08-03)


### Features

* split into two packages, use lerna as publisher ([#22](https://github.com/rx-ts/eslint-mdx/issues/22)) ([349b1ff](https://github.com/rx-ts/eslint-mdx/commit/349b1ff)), closes [#21](https://github.com/rx-ts/eslint-mdx/issues/21)

## [0.8.0](https://github.com/rx-ts/eslint-mdx/compare/v0.7.1...v0.8.0) (2019-08-03)


### Features

* split into two packages, use lerna as publisher ([#22](https://github.com/rx-ts/eslint-mdx/issues/22)) ([349b1ff](https://github.com/rx-ts/eslint-mdx/commit/349b1ff)), closes [#21](https://github.com/rx-ts/eslint-mdx/issues/21)

## [0.7.1](https://github.com/rx-ts/eslint-mdx/compare/v0.7.0...v0.7.1) (2019-08-03)


### Features

* export a parse function for compatibility usage ([0fdaf9f](https://github.com/rx-ts/eslint-mdx/commit/0fdaf9f))

## [0.7.0](https://github.com/rx-ts/eslint-mdx/compare/v0.6.0...v0.7.0) (2019-08-02)


### Features

* add new rule no-unescaped-entities ([dca8633](https://github.com/rx-ts/eslint-mdx/commit/dca8633))

## [0.6.0](https://github.com/rx-ts/eslint-mdx/compare/v0.5.0...v0.6.0) (2019-08-02)


### Features

* add new rule `no-jsx-html-comments`, close [#13](https://github.com/rx-ts/eslint-mdx/issues/13) ([67ba91e](https://github.com/rx-ts/eslint-mdx/commit/67ba91e))
* processor support for normalize comments in jsx [ci skip] ([114831c](https://github.com/rx-ts/eslint-mdx/commit/114831c))

## [0.5.0](https://github.com/rx-ts/eslint-mdx/compare/v0.4.1...v0.5.0) (2019-07-31)


### Features

* add `@rxts/mdx/no-unused-expressions` rule to replace eslint's ([5d62b55](https://github.com/rx-ts/eslint-mdx/commit/5d62b55))

## [0.4.1](https://github.com/rx-ts/eslint-mdx/compare/v0.4.0...v0.4.1) (2019-07-31)


### Features

* perf: extract `parseMdx`, exports everything in package ([1f965a1](https://github.com/rx-ts/eslint-mdx/commit/1f965a1))

## [0.4.0](https://github.com/rx-ts/eslint-mdx/compare/v0.3.1...v0.4.0) (2019-07-31)


### Features

* add body and comments into ast, fix [#12](https://github.com/rx-ts/eslint-mdx/issues/12) ([5b28bd5](https://github.com/rx-ts/eslint-mdx/commit/5b28bd5))

## [0.3.1](https://github.com/rx-ts/eslint-mdx/compare/v0.3.0...v0.3.1) (2019-07-30)


### Bug Fixes

* `.*` could not match multi lines, use `[\s\S]*` instead, close [#4](https://github.com/rx-ts/eslint-mdx/issues/4) ([f7e7efe](https://github.com/rx-ts/eslint-mdx/commit/f7e7efe))

## [0.3.0](https://github.com/rx-ts/eslint-mdx/compare/v0.2.1...v0.3.0) (2019-07-30)


### Bug Fixes

* inline jsx and comment parsing error, close [#4](https://github.com/rx-ts/eslint-mdx/issues/4), [#7](https://github.com/rx-ts/eslint-mdx/issues/7) ([5297a0b](https://github.com/rx-ts/eslint-mdx/commit/5297a0b))

## [0.2.1](https://github.com/rx-ts/eslint-mdx/compare/v0.2.0...v0.2.1) (2019-07-30)


### Bug Fixes

* eslint@5 is actually supported ([e470ddc](https://github.com/rx-ts/eslint-mdx/commit/e470ddc))

## [0.2.0](https://github.com/rx-ts/eslint-mdx/compare/v0.1.3...v0.2.0) (2019-07-30)


### Features

* support custom parser like babel-eslint ([b718574](https://github.com/rx-ts/eslint-mdx/commit/b718574))

## [0.1.3](https://github.com/rx-ts/eslint-mdx/compare/v0.1.2...v0.1.3) (2019-07-29)


### Bug Fixes

* upgrade peer dependency eslint to >= 6.0.0 ([#3](https://github.com/rx-ts/eslint-mdx/issues/3) [#4](https://github.com/rx-ts/eslint-mdx/issues/4)) ([f0ab288](https://github.com/rx-ts/eslint-mdx/commit/f0ab288))

## [0.1.2](https://github.com/rx-ts/eslint-mdx/compare/v0.1.1...v0.1.2) (2019-07-29)


### Bug Fixes

* show correct column and line on lint error ([90c5390](https://github.com/rx-ts/eslint-mdx/commit/90c5390))

## [0.1.1](https://github.com/rx-ts/eslint-mdx/compare/v0.1.0...v0.1.1) (2019-07-29)


### Bug Fixes

* overrides in node_modules seems not working ([d2f1535](https://github.com/rx-ts/eslint-mdx/commit/d2f1535))

## 0.1.0 (2019-07-29)


### Features

* traverse ast nodes ([6a71e25](https://github.com/rx-ts/eslint-mdx/commit/6a71e25))
