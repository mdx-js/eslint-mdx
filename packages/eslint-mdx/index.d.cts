import * as mdx from './lib/index.js'

export = mdx
