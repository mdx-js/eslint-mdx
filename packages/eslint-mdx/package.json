{
  "name": "eslint-mdx",
  "version": "1.6.4",
  "description": "ESLint Parser for MDX",
  "repository": "git+https://github.com/mdx-js/eslint-mdx.git",
  "homepage": "https://github.com/mdx-js/eslint-mdx/blob/master/packages/eslint-mdx",
  "funding": {
    "type": "opencollective",
    "url": "https://opencollective.com/unified"
  },
  "author": "JounQin <admin@1stg.me>",
  "license": "MIT",
  "engines": {
    "node": ">=7.0.0"
  },
  "main": "lib/cjs",
  "module": "lib",
  "es2015": "lib/es2015",
  "fesm5": "lib/esm",
  "types": "lib",
  "files": [
    "lib",
    "typings.d.ts"
  ],
  "keywords": [
    "eslint",
    "eslint-parser",
    "eslint-parser-mdx",
    "mdx",
    "eslint-mdx"
  ],
  "peerDependencies": {
    "eslint": ">=5.0.0"
  },
  "dependencies": {
    "espree": "^6.1.2",
    "remark-mdx": "^1.5.2",
    "remark-parse": "^7.0.2",
    "tslib": "^1.10.0",
    "unified": "^8.4.2"
  }
}
