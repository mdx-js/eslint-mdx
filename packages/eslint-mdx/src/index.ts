export * from './helpers'
export * from './parser'
export * from './regexp'
export * from './traverse'
export * from './types'
