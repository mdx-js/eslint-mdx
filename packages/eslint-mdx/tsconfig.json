{
  "extends": "../../tsconfig.lib",
  "compilerOptions": {
    "composite": true,
    "rootDir": "src",
    "outDir": "lib"
  },
  "include": ["src", "shim.d.ts"]
}
