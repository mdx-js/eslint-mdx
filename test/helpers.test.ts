import path from 'path'

import { arrayify, getPositionAt } from 'eslint-mdx'
import { getGlobals, getShortLang, requirePkg } from 'eslint-plugin-mdx'

describe('Helpers', () => {
  it('should arrayify items correctly', () => {
    expect(arrayify([])).toEqual([])
    expect(arrayify(1, 2)).toEqual([1, 2])
    expect(arrayify(1, 2, null)).toEqual([1, 2])
    expect(arrayify([1, 2], [1, 2])).toEqual([1, 2, 1, 2])
  })

  it('should short lang for filename correctly', () => {
    expect(getShortLang('1.Markdown')).toBe('md')
    expect(getShortLang('2.Markdown', false)).toBe('Markdown')
    expect(getShortLang('3.Markdown', { Markdown: 'mkdn' })).toBe('mkdn')
    expect(getShortLang('4.Markdown', { markdown: 'mkdn' })).toBe('mkdn')
  })

  it('should get correct loc range range', () => {
    const code = `
# Header

- jsx in list <div>
    <a href="link">content</a>
  </div>
    `.trim()
    expect(getPositionAt(code, code.indexOf('Header'))).toMatchSnapshot()
    expect(getPositionAt(code, code.indexOf('link'))).toMatchSnapshot()
    expect(getPositionAt(code, code.indexOf('content'))).toMatchSnapshot()
  })

  it('should resolve globals correctly', () => {
    expect(getGlobals({})).toEqual({})
    expect(getGlobals(['a', 'b'])).toEqual({
      a: false,
      b: false,
    })
  })

  it('should resolve package correctly', () => {
    expect(requirePkg('@1stg/config', 'commitlint')).toBeDefined()
    expect(requirePkg('lint', 'remark')).toBeDefined()
    expect(requirePkg('remark-parse', 'non existed')).toBeDefined()
    expect(
      requirePkg('./.eslintrc', 'non existed', path.resolve('package.json')),
    ).toBeDefined()
  })

  it('should throw on non existed package', () =>
    expect(() => requirePkg('@1stg/config', 'unexpected-')).toThrow())
})
