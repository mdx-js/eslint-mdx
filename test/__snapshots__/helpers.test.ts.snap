// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers should get correct loc range range 1`] = `
Object {
  "column": 2,
  "line": 1,
}
`;

exports[`Helpers should get correct loc range range 2`] = `
Object {
  "column": 13,
  "line": 4,
}
`;

exports[`Helpers should get correct loc range range 3`] = `
Object {
  "column": 19,
  "line": 4,
}
`;
